### PR TITLE
Add basic failover functionality

### PR DESCRIPTION
**Pull request by gtsonevv**
## Pre-flight checklist

- [x] I have read the [Contributing Guidelines on pull requests](https://github.com/near/near-api-js/blob/master/CONTRIBUTING.md).
- [x] Commit messages follow the [conventional commits](https://www.conventionalcommits.org/) spec
- [x] **If this is a code change**: I have written unit tests.
- [x] **If this changes code in a published package**: I have run `pnpm changeset` to create a `changeset` JSON document appropriate for this change.
- [x] **If this is a new API or substantial change**: the PR has an accompanying issue (closes [#733](https://github.com/near/near-api-js/issues/733)) and the maintainers have approved on my working plan.

## Motivation

Added support for multiple API Keys (since we can have multiple RPC Servers)
Added basic failover functionality
With this design of failover, we will change (rotate) our server when needed and stick to this new server for our next calls. It will help to avoid long responses when first (or more) servers are down.

For now, server rotation is happening only on Timeout error.

Also, check this comment: https://github.com/near/near-api-js/issues/733#issuecomment-998869162

## Test Plan

Added unit tests.

## Related issues/PRs

https://github.com/near/near-api-js/issues/733
https://github.com/near/near-api-js/pull/760

### CODE REVIEW

File: packages/cookbook/failover-configuration.js [hunk 1]
```diff
@@ -0,0 +1,36 @@
+// Demonstrates how to use failover functionality RPC Servers
+const { providers, connect } = require("near-api-js");
+
+const MAIN_RPC_SERVER = '<RPC Server 1>';
+const FAILOVER_RPC_SERVER_1 = '<RPC Server 2>';
+const FAILOVER_RPC_SERVER_2 = '<RPC Server 1>';
+
+// Provider example
+const provider = new providers.JsonRpcProvider({
+    // preoritized list of RPC Servers
```
Comment (vikinatora) [inline packages/cookbook/failover-configuration.js:10]:
```suggestion
    // prioritized list of RPC Servers
```

File: packages/providers/src/fetch_json.ts [hunk 2]
```diff
@@ -24,36 +26,45 @@ export async function fetchJson(connectionInfoOrUrl: string | ConnectionInfo, js
     } else {
         connectionInfo = connectionInfoOrUrl as ConnectionInfo;
     }
+    
+    if (Array.isArray(connectionInfo.url) && !connectionInfo.url.length) {
+        throw new Error('The prioritized array of RPC Server URLs should not be empty');
```
Comment (vikinatora) [inline packages/providers/src/fetch_json.ts:31]:
```suggestion
        throw new Error('The prioritized array of RPC Server URLs must not be empty');
```

File: packages/providers/src/fetch_json.ts [hunk 3]
```diff
@@ -24,36 +26,45 @@ export async function fetchJson(connectionInfoOrUrl: string | ConnectionInfo, js
...
         try {
 
-            const response = await (global.fetch ?? (await import('./fetch')).default)(connectionInfo.url, {
+            const response = await (global.fetch ?? (await import('./fetch')).default)(currentRpcServer, {
                 method: json ? 'POST' : 'GET',
                 body: json ? json : undefined,
-                headers: { ...connectionInfo.headers, 'Content-Type': 'application/json' }
+                headers: {
+                    'Content-Type': 'application/json',
+                    'x-api-key': connectionInfo.apiKeys ? connectionInfo.apiKeys[currentRpcServer] : undefined,
```
Comment (vikinatora) [inline packages/providers/src/fetch_json.ts:43]: I think that `null` is semantically better than `undefined` in this case
```suggestion
                    'x-api-key': connectionInfo.apiKeys ? connectionInfo.apiKeys[currentRpcServer] : null,
```

File: packages/providers/src/json-rpc-provider.ts [hunk 4]
```diff
@@ -333,6 +333,19 @@ export class JsonRpcProvider extends Provider {
         return await this.sendJsonRpc('gas_price', [blockId]);
     }
 
+    /**
+     * Part of the RPC failover design.
+     * Changing current (first) rpc server and moves it to the end of the queue.
```
Comment (vikinatora) [inline packages/providers/src/json-rpc-provider.ts:338]:
```suggestion
     * Rotates the current RPC server with a new one (if available) and moves the former to the end of the queue.
```

File: packages/providers/src/json-rpc-provider.ts [hunk 5]
```diff
@@ -333,6 +333,19 @@ export class JsonRpcProvider extends Provider {
...
+     */
+    private rotateRpcServers() {
+        if (typeof this.connection.url === 'string') { return; }
```
Comment (vikinatora) [inline packages/providers/src/json-rpc-provider.ts:341]:
```suggestion
        if (typeof this.connection.url === 'string' || this.connection.url.length === 1) { return; }
```

File: packages/providers/test/providers.test.js [hunk 6]
```diff
@@ -187,3 +187,46 @@ test('near json rpc fetch node status', async () => {
     let response = await provider.status();
     expect(response.chain_id).toBeTruthy();
 });
+
+test('near json rpc url - empty array', async () => {
+    const provider = new JsonRpcProvider({ url: [] });
+    try {
+        await provider.status();
+    } catch (e) {
+        expect(e.message).toEqual('The prioritized array of RPC Server URLs should not be empty');
```
Comment (vikinatora) [inline packages/providers/test/providers.test.js:196]:
```suggestion
        expect(e.message).toEqual('The prioritized array of RPC Server URLs cannot be empty');
```

**Review by vikinatora — COMMENTED** [review]
Looks good, I've suggested some minor changes

**Comment by denbite** [comment]
The failover mechanism looks exciting to me, I've been tracking it since #760, and it's going to be very useful

However, I wanted to make sure the implementation would cover all use cases, and one of them is what if a user has different connection data for providers

For instance, how can the user provide the following list of RPCs?

```
[
    {
        "url": "https://rpc.mainnet.near.org",
        "timeout": 2500
    },
    {
        "url": "https://another-rpc.cloud.com",
        "timeout": 2000,
        "headers": {
            "X-Api-Key": "some-string-key"
        }
    },
    {
        "url": "https://one-more-rpc.cloud2.com",
        "headers": {
            "Authorization": "Bearer jwt.token"
        }
    }
]
```

And it's widespread for each provider to have its own set of headers or other parameters that are required, there's no standard for it

I propose not to extend the `JsonRpcProvider` class because its primary responsibility is to know how to communicate with RPC and not load-balancing providers, but instead implement the new provider class, let's say `FailoverJsonRpcProvider` which is going to be the only place that would know about algorithms that are used for failover

```
export class FailoverJsonRpcProvider extends Provider {
  constructor(
    readonly providers: JsonRpcProvider[],
    // parameters to customize the failover mechanism (retries, total time, etc)
    readonly failoverConfiguration: any
  ) {}


  ...
}
```

To make it work properly, we would also need to introduce the new field `nodeConnections` in the configuration
```
export interface NearConfig {
    ...
    nodeConnections: ConnectionInfo | ConnectionInfo[],
    ...
}
```

@gtsonevv 
Please leave your thoughts about this proposal, I want it to be considered before merging PR

**Comment by vikinatora** [comment]
Hey @denbite, thank you for the feedback. On first glance your approach seems reasonable, we'll thoroughly analyze it internally and get back to you.

**Comment by frol** [comment]
It would be great to have the link to the available RPC providers also easily discoverable for the devs configuring the connection: https://docs.near.org/api/rpc/providers

**Comment by denbite** [comment]
quite a few people have been asking me recently about the failover mechanism to make their applications more robust, @vikinatora @gtsonevv do you guys have ETA for this feature?

P.S. let me know if you need any help in developing, I'd be happy to take over and deliver this great functionality to active developers as quickly as possible

**Comment by vikinatora** [comment]
@denbite We can give you an ETA, as we have commited to delivering other functionalities. We'd be happy for you take over and finish this functionality.

**Comment by vikinatora** [comment]
@gtsonevv, we can close this one.

**Comment by gagdiez** [comment]
closing as requested by @vikinatora

**Pull request closed without merging**